contracts/Gateway/NativeTokenGateway.sol

```diff
@@ -0,0 +1,200 @@
+     */
+    constructor(IVToken vWrappedNativeToken) {
+        ensureNonzeroAddress(address(vWrappedNativeToken));
+
+        vWNativeToken = vWrappedNativeToken;
+        wNativeToken = IWrappedNative(vWNativeToken.underlying());
+    }
+
+    /**
+     * @notice To receive Native when msg.data is empty
+     */
+    receive() external payable {}
+
+    /**
+     * @notice To receive Native when msg.data is not empty
+     */
+    fallback() external payable {}
+
+    /**
+     * @notice Wrap Native, get wNativeToken, mint vWNativeToken, and supply to the market.
+     * @param minter The address on behalf of whom the supply is performed.
+     * @custom:error ZeroAddressNotAllowed is thrown if address of minter is zero address
+     * @custom:error ZeroValueNotAllowed is thrown if mintAmount is zero
+     * @custom:event TokensWrappedAndSupplied is emitted when assets are supplied to the market
+     */
+    function wrapAndSupply(address minter) external payable nonReentrant {
+        ensureNonzeroAddress(minter);
+
+        uint256 mintAmount = msg.value;
+        ensureNonzeroValue(mintAmount);
+
+        wNativeToken.deposit{ value: mintAmount }();
+        IERC20(address(wNativeToken)).safeApprove(address(vWNativeToken), mintAmount);
+
+        vWNativeToken.mintBehalf(minter, mintAmount);
+
+        IERC20(address(wNativeToken)).safeApprove(address(vWNativeToken), 0);
+        emit TokensWrappedAndSupplied(minter, address(vWNativeToken), mintAmount);
+    }
+
+    /**
+     * @notice Redeem vWNativeToken, unwrap to Native Token, and send to the user
+     * @param redeemAmount The amount of underlying tokens to redeem
+     * @custom:error ZeroValueNotAllowed is thrown if redeemAmount is zero
+     * @custom:event TokensRedeemedAndUnwrapped is emitted when assets are redeemed from a market and unwrapped
+     */
+    function redeemUnderlyingAndUnwrap(uint256 redeemAmount) external nonReentrant {
+        ensureNonzeroValue(redeemAmount);
+
+        uint256 balanceBefore = wNativeToken.balanceOf(address(this));
+        vWNativeToken.redeemUnderlyingBehalf(msg.sender, redeemAmount);
+        uint256 balanceAfter = wNativeToken.balanceOf(address(this));
+
+        uint256 redeemedAmount = balanceAfter - balanceBefore;
+        wNativeToken.withdraw(redeemedAmount);
+
+        _safeTransferNativeTokens(msg.sender, redeemedAmount);
+        emit TokensRedeemedAndUnwrapped(msg.sender, address(vWNativeToken), redeemedAmount);
+    }
+
+    /**
+     * @dev Borrow wNativeToken, unwrap to Native, and send to the user
+     * @param borrowAmount The amount of underlying tokens to borrow
+     * @custom:error ZeroValueNotAllowed is thrown if borrowAmount is zero
+     * @custom:event TokensBorrowedAndUnwrapped is emitted when assets are borrowed from a market and unwrapped
+     */
+    function borrowAndUnwrap(uint256 borrowAmount) external nonReentrant {
+        ensureNonzeroAddress(address(vWNativeToken));
```

Comment:
I don't think this check is necessary

Reply:
https://github.com/VenusProtocol/isolated-pools/pull/361/commits/bb127763e9a97d1e50427a7a247888c1b2cbfeae